### PR TITLE
PIL-1024: exchange allowances per account

### PR DESCRIPTION
**Pull request by poocart**
https://linear.app/pillarproject/issue/PIL-1024/exchange-allowances

Adds exchange allowances per account to support recently adde Etherspot account.

### CODE REVIEW

File: src/models/Offer.js [hunk 1]
```diff
@@ -70,18 +62,22 @@ export type OfferOrder = {
...
 export type Allowance = {
   provider: string,
   fromAssetCode: string,
   toAssetCode: string,
   transactionHash: string,
   enabled?: boolean,
-}
+};
+
+export type Allowances = {
```
Comment (mdjastrzebski) [inline src/models/Offer.js:81]: Consider renaming to `AccountAllowances` or `AllowancesPerAccount` for readability.

File: src/actions/exchangeActions.js [hunk 2]
```diff
@@ -328,29 +370,31 @@ export const checkEnableExchangeAllowanceTransactionsAction = () => {
...
-            emoji: 'handshake',
-            autoClose: true,
-          });
-          dispatch(enableExchangeAllowanceByHashAction(transactionHash));
-        }
-      });
+    Object.keys(exchangeAllowances).map((accountId) =>
+      exchangeAllowances[accountId]
+        .filter(({ enabled }) => !enabled)
+        .map(({ transactionHash, fromAssetCode, toAssetCode }) => {  // eslint-disable-line
```
Comment (mdjastrzebski) [inline src/actions/exchangeActions.js:376]: Looks like this should be 'forEach'

```suggestion
        .forEach(({ transactionHash, fromAssetCode, toAssetCode }) => { 
```

File: src/actions/exchangeActions.js [hunk 3]
```diff
@@ -263,49 +270,84 @@ export const addExchangeAllowanceIfNeededAction = ({
...
-    const { exchange: { data: { allowances: _allowances = [] } } } = getState();
-    const allowance = _allowances.find(
+export const enableExchangeAllowanceByHashAction = (transactionHash: string, senderAccountId: string) => {
+  return (dispatch: Dispatch, getState: GetState) => {
+    const allowances = allAccountsAllowancesSelector(getState());
+
+    const combinedAllowances = Object.keys(allowances).reduce((combined, accountId) => [
+      ...combined,
+      ...allowances[accountId],
+    ], []);
```
Comment (mdjastrzebski) [inline src/actions/exchangeActions.js:315]: This should be the same, but more idiomatic.
```suggestion
    const combinedAllowances = Object.keys(allowances).flatMap((accountId) => allowances[accountId]);
```

File: src/actions/exchangeActions.js [hunk 4]
```diff
@@ -263,49 +270,84 @@ export const addExchangeAllowanceIfNeededAction = ({
...
     });
-    const allowances = _allowances
-      .filter(
-        ({ transactionHash: _transactionHash }) => _transactionHash !== transactionHash,
-      );
-    allowances.push(updatedAllowance);
-    dispatch(saveDbAction('exchangeAllowances', { allowances }, true));
+
+    const updatedAllowances = Object.keys(allowances).reduce((updated, allowancesAccountId) => {
+      if (senderAccountId !== allowancesAccountId) return updated;
```
Comment (mdjastrzebski) [inline src/actions/exchangeActions.js:334]: This part seems a bit weird. So for all entries in `allowances` map we are interested only in allowanes for `senderAccountId`. IMO it looks like there is no need to use `reduce` then, but just get the `allowances[senderAccountId]` and do the same thing as in `reduce` body.

WDYT?

File: src/actions/exchangeActions.js [hunk 5]
```diff
@@ -263,49 +270,84 @@ export const addExchangeAllowanceIfNeededAction = ({
...
     const updatedAllowance = {
       ...allowance,
       enabled: true,
     };
+
     dispatch({
-      type: UPDATE_EXCHANGE_ALLOWANCE,
-      payload: updatedAllowance,
+      type: UPDATE_ACCOUNT_EXCHANGE_ALLOWANCE,
+      payload: { accountId: senderAccountId, allowance: updatedAllowance },
```
Comment (mdjastrzebski) [inline src/actions/exchangeActions.js:330]: Seems like we could be duplicating given allowance entry from one account to another, is that the intent?

**Review by mdjastrzebski — APPROVED** [review]
Overall looks good, but there are some minor details that require attention.

File: src/actions/exchangeActions.js (hunk 5 above)
Comment (poocart) [reply src/actions/exchangeActions.js:330]: Yes. Allowance is set per each sender account.

File: src/services/dataMigration/exchangeAllowances.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+import { isArray } from 'lodash';
+import { getAccountId, isArchanovaAccount } from 'utils/accounts';
+
+export default function (storageData: Object) {
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/services/dataMigration/exchangeAllowances.js:24]: I think for the matter on this PR let's leave as it is.

File: src/actions/exchangeActions.js (hunk 2 above)
Comment (poocart) [reply src/actions/exchangeActions.js:376]: Old code, didn't bother to fix as it should have been quick fix, but will change, thanks 👍

File: src/reducers/exchangeReducer.js [hunk 7]
```diff
@@ -141,32 +150,31 @@ export default function exchangeReducer(
...
-          allowances: [
+          allowances: {
             ...state.data.allowances,
-            action.payload,
-          ],
+            [accountId]: [...accountAllowances, allowance],
+          },
           hasNotification: true,
         },
       };
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/reducers/exchangeReducer.js:166]: Technically we could, but we have lint rule that forbids redeclaring same name vars under single switch.

File: src/actions/exchangeActions.js (hunk 4 above)
Comment (poocart) [reply src/actions/exchangeActions.js:334]: You're absolutely right, this ticket is little rush of moving existing pieces, changed, thanks, good catch 👍

**Pull request merged**